crates/ide-completion/src/render.rs

```diff
@@ -638,20 +639,31 @@ fn compute_exact_name_match(ctx: &CompletionContext<'_>, completion_name: &str)
 fn compute_ref_match(
     ctx: &CompletionContext<'_>,
     completion_ty: &hir::Type,
-) -> Option<hir::Mutability> {
+) -> Option<CompletionItemRefMode> {
     let expected_type = ctx.expected_type.as_ref()?;
     if completion_ty != expected_type {
-        let expected_type_without_ref = expected_type.remove_ref()?;
-        if completion_ty.autoderef(ctx.db).any(|deref_ty| deref_ty == expected_type_without_ref) {
-            cov_mark::hit!(suggest_ref);
-            let mutability = if expected_type.is_mutable_reference() {
-                hir::Mutability::Mut
-            } else {
-                hir::Mutability::Shared
-            };
-            return Some(mutability);
-        };
+        if let Some(expected_type_without_ref) = expected_type.remove_ref() {
+            if completion_ty.autoderef(ctx.db).any(|deref_ty| deref_ty == expected_type_without_ref)
+            {
+                cov_mark::hit!(suggest_ref);
+                let mutability = if expected_type.is_mutable_reference() {
+                    hir::Mutability::Mut
+                } else {
+                    hir::Mutability::Shared
+                };
+                return Some(CompletionItemRefMode::Reference(mutability));
+            }
+        }
+    }
+
+    if let Some(completion_ty_without_ref) = completion_ty.remove_ref() {
```

Comment:
We can move this to the start of the function as `let expected_type_without_ref = expected_type.remove_ref()?;` since the branch above makes use of it as well either way to deduplicate a bit

Reply:
Agreed I'll make the change

Reply:
@Veykril I pushed the update to move both of these methods to the top of the function. Let me know if this is not what you were looking for.